src/modules/wreck/job.c

```diff
@@ -543,74 +545,184 @@ static int flux_attr_get_int (flux_t *h, const char *attr, int *valp)
+        if (strcmp (stream, "stderr") == 0)
+            level = LOG_ERR;
+        flux_log (h, level, "job%ju: wrexecd says: %s",
+                            (uintmax_t) job->id,
+                            (char *) data);
+    }
+    free (data);
+    return;
+}
 
-    if ((sid = setsid ()) < 0)
-        fprintf (stderr, "setsid: %s\n", strerror (errno));
+static void cmb_exec_cb (flux_future_t *f, void *arg)
+{
+    int64_t pid = 0;
+    const char *type = NULL;
+    const char *state = NULL;
+    const flux_msg_t *msg;
+    flux_t *h = flux_future_get_flux (f);
+    struct wreck_job *job = arg;
 
-    /*
-     *  NOTE: There used to be a double fork here, presumably to
-     *  "daemonize" wrexecd, however I'm not sure that is warranted
-     *   nor even advisable. With the setsid above, the wrexecd
-     *   process should be reparented to init.
-     */
+    if (flux_future_get (f, &msg) < 0) {
+        flux_log_error (h, "cmb_exec_cb: flux_future_get");
+        flux_future_destroy (f);
+        return;
+    }
+    if (flux_msg_unpack (msg, "{s?s,s?s,s:i}",
+                              "type", &type, "state", &state,
+                              "pid", &pid) < 0) {
+        flux_log_error (h, "cmb_exec_cb: flux_msg_unpack");
+        flux_future_destroy (f);
+    }
 
-    fdwalk (exec_close_fd, NULL);
-    if (setenv ("FLUX_URI", local_uri, 1) < 0)
-        fprintf (stderr, "setenv: %s\n", strerror (errno));
-    else if (execvp (av[0], av) < 0)
-        fprintf (stderr, "wrexecd exec: %s\n", strerror (errno));
-    exit (255);
+    if (type && strcmp (type, "io") == 0)
+        spawn_io_cb (h, job, msg);
+    else if (strcmp (state, "Exited") == 0) {
+        int status = 1;
+        if (flux_msg_unpack (msg, "{s:i}", "status", &status) < 0)
+            flux_log_error (h, "spawn response: failed to unpack exit status");
+        if (WIFSIGNALED (status))
+            flux_log_error (h, "job%ju: wrexecd: %s",
+                                (uintmax_t) job->id,
+                                strsignal (WTERMSIG (status)));
+        else if (WEXITSTATUS (status) != 0)
+            flux_log_error (h, "job%ju: wrexecd: Exit %d",
+                                (uintmax_t) job->id, WEXITSTATUS (status));
+
+        wreck_job_destroy (job);
+        /* Done with this job, it is safe to destroy future */
+        flux_future_destroy (f);
+        return;
+    }
+    else flux_log (h, LOG_ERR, "job%ju: unknown state %s",
+                               (uintmax_t) job->id, state);
+    flux_future_reset (f);
+    return;
 }
 
-static int spawn_exec_handler (flux_t *h, struct wreck_job *job)
+static json_t *wrexecd_cmdline_create (flux_t *h, struct wreck_job *job)
 {
-    pid_t pid;
+    json_t *o;
     const char *wrexecd_path;
+    char buf [4096];
+    int n;
 
     if (!(wrexecd_path = flux_attr_get (h, "wrexec.wrexecd_path", NULL))) {
         flux_log_error (h, "spawn_exec_handler: flux_attr_get");
```

Comment:
spawn_exec_handler -> wrexecd_cmdline_create? (or use `__FUNCTION__`)

Reply:
I really dislike using `__FUNCTION__` because you can't effectively search code based on the actual error message. I'll update the error text if that is required, but the error is during spawn_exec...

Reply:
Sorry - reasonable. Maybe I should have said the error prefix in this message is different than the one right after it.  And it's minor so not worth making a big deal over.

Reply:
Oh, yeah, the inconsistency is a good catch and it should be fixed.